Integrate LNDg to Raspiblitz

allyourbankarebelongtous opened the pull request:
Integrates LNDg (popular auto-rebalancer and auto-fee setting node manager) to the Raspiblitz via the SERVICES menu. bonus.lndg.sh runs a complete install of LNDg, including an nginx server and systemd backend services. bonus.lndg.sh along with included .conf files in /home/admin/assets/nginx/sites-available/ includes support for ssl and ssl over tor similar to other Raspiblitz install scripts. bonus.lndg.sh also moves the LNDg database to /mnt/hdd/app-data/lndg and simlinks it for a persistent database over updates, and _provision_.sh combined with bonus.lndg.sh ensures that the update will not change LNDg. bonus.lndg.sh automatically includes tor if tor is set in raspiblitz.conf. bonus.lndg.sh installs LNDg in /home/lndg/lndg under user lndg. Also includes the ability to update LNDg from the UPDATE menu, and includes an INFO menu when LNDg is installed, similar to other Raspiblitz services.

Fix Required: Currently changes the way python is called, resulting in other scripts breaking (see comment below). Working on fixing this. -- FIXED

TODO: 
Future: Integrate signature check if NGU ever starts signing his LNDg updates. 
Future: Enable bonus.lndg.sh to update to a particular version instead of just the latest version.
Working on: Give users the option of keeping or deleting database (currently always keeps database and always uses it on next install). --DONE
Working on: Integrate into blitzapi
Working on: Allow users to pick username on install

allyourbankarebelongtous commented:
After further checking, there seems to be a problem with the install. The install breaks other instances on the Raspiblitz that call scripts using "python" instead of "python3". See [issue 3003](https://github.com/rootzoll/raspiblitz/issues/3003). For example, my blitzapi.service didn't work until I edited the .service file from "/usr/bin/python -m uvicorn app.main:app --port 11111 --host=0.0.0.0 --root-path /api" to "/usr/bin/python3 -m uvicorn app.main:app --port 11111 --host=0.0.0.0 --root-path /api". It seems that installing the method for installing dependencies for the uwsgi server makes "python" not automatically python3. This is the same issue as issue 3003 (I had installed lndg at the time using a similar method). Help would be appreciated.

rootzoll commented:
Thanks for the PR - marked to be part of v1.8.1 release.

FYI best to base PRs on `dev` branch to keep up with other changed happening.

rootzoll commented:
Note: Linked to issue #3230

allyourbankarebelongtous commented:
Will do. I found a fix for the python issue as well. Switched from uwsgi to gunicorn for app server, runs only on python 3 with no dependencies. I'll recommit using dev branch with new changes.

allyourbankarebelongtous commented:
Ok, this commit fixes the issue with python by using gunicorn instead of uwsgi as the application server. Tests good, seems to be super stable with low overhead. Runs using systemd and installs with no dependencies. This commit also allows users to either delete or keep the database on uninstall, and allows users to use an existing database (if one exists) or delete it and restart from scratch on install. Tests good, feedback welcome.

MEdGalo commented:
Hi, I seems that after the installation of lndg  it the blocks the shutdown of LND when using the reboot or shutdown option from the raspiblitz menu. Could you please check if this indeed the case?

allyourbankarebelongtous commented:
I'll run some troubleshooting over the weekend if I can, but it does appear to do that. May require a modification to the shutdown/reboot scripts.

allyourbankarebelongtous commented:
New push fixing shutdown script. Checks if LNDg is installed and if so, shuts down services prior to shutdown of raspiblitz. Tests good from my end.

MEdGalo commented:
This fixed it for me as well.

MEdGalo commented:
My failed HTLCs page doesn't seem to update, does it work for you?

allyourbankarebelongtous commented:
> My failed HTLCs page doesn't seem to update, does it work for you?

Works for me, try `sudo systemctl status htlc-stream-service` and see what it says.

MEdGalo commented:
htlc-stream-lndg.service - Run HTLC Stream For Lndg
     Loaded: loaded (/etc/systemd/system/htlc-stream-lndg.service; enabled; vendor preset: enabled)
     Active: inactive (dead)

I started it now with sudo systemctl start  htlc-stream-lndg.service, but I don't know then why it didn't start in the first place

allyourbankarebelongtous commented:
Can you paste the content of the service file here?
`sudo nano /etc/systemd/system/htlc-stream-lndg.service` and copy/paste?

allyourbankarebelongtous commented:
Looks good. I know the script starts it when it's installed and I haven't had mine not start again on boot. If it becomes a persistent issue let me know.

rootzoll commented:
So this really looks good - install tests ran fine :) Lets add it in the next release (called 1.9.0) for SSH install.

TODO: CHANGES.md & README.md (in the listing of available apps)

allyourbankarebelongtous commented:
Thanks for the feedback! Do you want me to update CHANGES.md and README.md? Anything else?

rootzoll commented:
CHANGES.md and README if you like - I can do too. The "passwordB" (see review above) would be a missing solution to implement.

allyourbankarebelongtous commented:
> CHANGES.md and README if you like - I can do too. The "passwordB" (see review above) would be a missing solution to implement.

PasswordB is already used. Don't see any issues with it in the reviews/comments, am I missing something?

I'll hack on CHANGES.md and README.md over the weekend.

allyourbankarebelongtous commented:
> > PasswordB is already used. Don't see any issues with it in the reviews/comments, am I missing something?
> 
> Again what I mean is how to change the admin password to lndg by command line? Can I get a `bonus.lndg.sh setpassword [NEWPASSWORD]` which will set a new password after install?
> 
> Also I dont think we should offer to change the login username - like printed in info menu `To change username, login using default username and go to: http://192.168.178.61:8889/lndg-admin` .. later the WebUI wants to show the user info like: "Use usernam 'lndg-admin' and your PasswordB" .. but if the user can change the admin user name this is the wrong information than.

I'll create a script that changes the password and edit the menu to just say "login with username lndg-admin and passwordB". New commit incoming this weekend!

Note: because lndg uses a Django interface, people will still be able to change their username and password by going to http://192.168.178.61:8889/lndg-admin, but the menu won't suggest it. I did try to find a way to get the username so the new username is always displayed by the menu but the username is only stored encrypted in the database, so I couldn't figure out how to do that.

allyourbankarebelongtous commented:
Ok, this last commit adds the ability to change the password using the script with `bonus.lndg.sh setpassword mynewpasswordB`. 

Note: To do this I had to create a new python file: the install script creates it when you install it, so if you have already installed and want to test this feature, either manually create the changepassword.py file or uninstall and reinstall LNDg using this updated script.

Also fixed the menu so it doesn't suggest changing the username and password via the Django web interface.

I don't have the most up to date CHANGES.md and README.md, so I'll defer to Rootzoll to update those, but I'm happy to provide inputs if needed.

allyourbankarebelongtous commented:
On further testing, I found that uninstalling and reinstalling using the same database resulted in an error due to an update that LNDg pushed. My database was v1.3.1, and LNDg was updated to v1.4.0, so when I reinstalled it installed v1.4.0 on top of an older database. My script didn't account for that, so I had to run the update script to migrate the changes. I added a migrate line to any install that uses an existing database, which should fix that.

I also modified the setpassword script to check for changepassword.py, and create it if it doesn't exist, making this work out of the box without an uninstall/reinstall to create the file. It still creates the file on install. 

Tests good.

allyourbankarebelongtous commented:
One more commit that fixes an issue that was discovered by another user attempting to update. I missed this because I uninstalled and reinstalled instead of updating through the update menu. 

This latest LNDg update made changes to settings.py which is auto-generated on initializing the app. This change makes it so that all updates will reinitialize settings.py in case future updates also require it.

Tests good.

rootzoll commented:
Ok will merge into `3313-lndg` feature branch and continue dev on this issue:
https://github.com/rootzoll/raspiblitz/issues/3230#issuecomment-1344516763

allyourbankarebelongtous commented on home.admin/config.scripts/bonus.lndg.sh:
```diff
@@ -0,0 +1,380 @@
+    sudo -u lndg virtualenv -p python3 .venv
+    sudo -u lndg .venv/bin/pip install -r requirements.txt
+    PASSWORD_B=$(sudo cat /mnt/hdd/bitcoin/bitcoin.conf | grep rpcpassword | cut -c 13-)
+    sudo -u lndg .venv/bin/python initialize.py -pw $PASSWORD_B
```
Is it too late for me to work it into the blitz.passwords.sh? I can submit a commit for that tonight if not.

rootzoll replied:
All good - did the password change already in the branch `3313-lndg`.

allyourbankarebelongtous replied:
Awesome, thanks! First PR ever for me.
